software/contrib/consequencer.py

```diff
@@ -0,0 +1,430 @@
+        
+        self.generateNewRandomCVPattern()
+
+        # Triggered when button 2 is released.
+        # Short press: Generate random CV for cv4-6
+        # Long press: Change operating mode
+        @b2.handler_falling
+        def b2Pressed():
+            
+            if ticks_diff(ticks_ms(), b2.last_pressed()) >  300:
+                if self.analogInputMode < 3:
+                    self.analogInputMode += 1
+                else:
+                    self.analogInputMode = 1
+            else:
+                # Move to next cv pattern if one already exists, otherwise create a new one
+                self.CvPattern += 1
+                if self.CvPattern == len(self.random4):
+                    self.generateNewRandomCVPattern()
+            
+        # Triggered when button 1 is released
+        # Short press: Play previous CV pattern for cv4-6
+        # Long press: Toggle random high-hat mode
+        @b1.handler_falling
+        def b1Pressed():
+            if ticks_diff(ticks_ms(), b1.last_pressed()) >  300:
+                self.random_HH = not self.random_HH
+            else:
+                # Play previous CV Pattern, unless we are at the first pattern
+                if self.CvPattern != 0:
+                    self.CvPattern -= 1
+
+        # Triggered on each clock into digital input. Output triggers.
+        @din.handler
+        def clockTrigger():
+            
+            if self.clock_step % self.clock_division == 0:
+
+                self.step_length = len(self.BD[self.pattern])
+                
+                # A pattern was selected which is shorter than the current step. Set to zero to avoid an error
+                if self.step >= self.step_length:
+                    self.step = 0 
+
+                # Set cv4-6 voltage outputs based on previously generated random pattern
+                cv4.voltage(self.random4[self.CvPattern][self.step])
+                cv5.voltage(self.random5[self.CvPattern][self.step])
+                cv6.voltage(self.random6[self.CvPattern][self.step])
+
+                # How much randomness to add to cv1-3
+                # As the randomness value gets higher, the chance of a randomly selected int being lower gets higher
+                if randint(0,99) < self.randomness:
+                    cv1.value(randint(0, 1))
+                    cv2.value(randint(0, 1))
+                    cv3.value(randint(0, 1))
+                else:
+                    cv1.value(int(self.BD[self.pattern][self.step]))
+                    cv2.value(int(self.SN[self.pattern][self.step]))                    
+
+                    # If randomize HH is ON:
+                    if self.random_HH:
+                        cv3.value(randint(0, 1))
+                    else:
+                        cv3.value(int(self.HH[self.pattern][self.step]))
+
+            # Reset clock step at 128 to avoid an overflow if running for a long time
+            if self.clock_step < 128:
```

Comment:
nit: Technically this is actually unnecessary because integers do not overflow in Python3. I'd also suggest removing this check because it could cause patterns with lengths not evenly divisible by 128 to skip a beat when this resets. So just keep a running counter going and Python do its thing. :)

https://docs.python.org/3/c-api/long.html#integer-objects

Reply:
Also I realized the default clock_division is 1 so you could also remove this and the block above can be unindented.

Reply:
>it could cause patterns with lengths not evenly divisible by 128 to skip a beat when this resets.

I'm wrong here. I meant `clock_divisions` evenly divisible by 128, not steps, but they just trigger the step, not advance the step, so that concern is moot.